core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala

```diff
@@ -244,6 +244,31 @@ class SparkHadoopUtil extends Logging {
   }
 
   /**
+   * List directories/files matching the path and return the `FileStatus` results.
+   * If the pattern is not a regexp then a simple `getFileStatus(pattern)`
+   * is called to get the status of that path.
+   * If the path/pattern does not match anything in the filesystem,
+   * an empty sequence is returned.
+   * @param pattern pattern
+   * @return a possibly empty array of FileStatus entries
+   */
+  def globToFileStatus(pattern: Path): Array[FileStatus] = {
```

Comment:
essentially if anything which might be a wildcard is hit, it gets handed off to the globber for the full interpretation. Same for ^ and ], which are only part of a pattern within the context of an opening [

Its only those strings which can be verified to be regexp free in a simple context-free string scan that say "absolutely no patterns here"

regarding the bigger change: most of it is isolation of the sensitive code *and the tests to verify behaviour*